### ISSUE TITLE
How to debug nuxt in JetBrains WebStorm?

### ISSUE THREAD

**Issue by xiaoyun-studio** [nuxt/nuxt.js#1577]
How to debug nuxt in JetBrains WebStorm?

**Comment by NicoPennec** [comment]
I'm not user of WebStorm, but you should be able to add this task to your package. json

```
"scripts" {
  "dev":"nuxt",
  "dev-debug":"node --debug node_modules/. bin/nuxt",
```

then run the `dev-debug` script via webstorm debug mode.

(inspired by VS Code debugging: https://medium.com/@marshallswain/debugging-nuxt-js-with-visual-studio-code-724920140b8f)

**Comment by awronski** [comment]
No sure how to do this.

I added `"dev-debug": "node --inspect-brk node_modules/.bin/nuxt"` to scripts.

Then configured IntelliJ like this:
![image](https://user-images.githubusercontent.com/1973725/30052730-c6281014-9226-11e7-8d77-e9c9f691d311.png)

Then run "Debug" and got an output:
```
"C:\dev\IntelliJ IDEA 172.2827.15\bin\runnerw.exe" C:\dev\nodejs\node.exe C:\dev\nodejs\node_modules\npm\bin\npm-cli.js run dev-debug --scripts-prepend-node-path=auto

> magpie-admin@1.0.0 dev-debug C:\dev\git\magpie2\magpie-admin
> node --inspect-brk node_modules/.bin/nuxt

Debugger listening on ws://127.0.0.1:9229/bf8465e4-2b15-4153-ade6-55d7e884e2a6
For help see https://nodejs.org/en/docs/inspector
```

I cannot access `ws://127.0.0.1:9229...` url.

What should I do?

**Comment by xiaoyunStudio** [comment]
debug config:
![1](https://user-images.githubusercontent.com/19432652/30055778-1271093a-9263-11e7-85c0-2c8bf4d67707.PNG)

add debugger to vue:
![2](https://user-images.githubusercontent.com/19432652/30055845-42341ef0-9263-11e7-9c5a-fcae0c579c62.PNG)

debug:
![3](https://user-images.githubusercontent.com/19432652/30055867-4c6a72a2-9263-11e7-8141-6c4cea17ecc2.PNG)

**Comment by awronski** [comment]
@xiaoyunStudio Thanks for help. The IDE starts the app, dosn't stops at breakpoints but at least stops at `debugger`.

But the problem is that all variables are undefined.
![image](https://user-images.githubusercontent.com/1973725/30058151-00e79e64-923a-11e7-9cb1-741f172206ee.png)

Any suggestions?

**Comment by AdmiralPotato** [comment]
I would also love to be able to use breakpoints. Anyone have any ideas?

The above setup using the `debugger` trick is also working for me in IntelliJ IDEA Ultimate. Thank you for sharing this, @xiaoyunStudio - even this is already helping a lot.

I have found it useful to wrap the `debugger` statement like this, so the browser will keep re-triggering the page load on changes, without getting stuck in the chrome devtools when it hits the broken line:
```javascript
if (context.isServer) {
    debugger
}
```

**Comment by pi0** [comment]
Wow. I really love this topic. Would be awesome if we have a section in docs FAQ about "How to debug SSR" before closing issue. (& Certainly, contributions welcome)

**Comment by awronski** [comment]
Hi! There must be a way to do this. WebStorm is the one or maybe event the best IDE. If it is possible to debug [React](https://blog.jetbrains.com/webstorm/2017/01/debugging-react-apps/) or [webpack](https://blog.jetbrains.com/webstorm/2015/09/debugging-webpack-applications-in-webstorm/) why not Nuxt.

I follow both @xiaoyunStudio suggestion and official WebStorm [docs](https://www.jetbrains.com/help/webstorm/run-debug-configuration-node-js.html) but it doesn't work as it suppose to. Especially the `undefined` variables are big problem. With this debuging is useless.

There must be someone who know how to do this :)

**Comment by alterhu2020** [comment]
Also same question for this debug feature , all the variables are `undefined` ,
Why close this topic here?

**Comment by fbrbovic** [comment]
I am also looking for answer to this, has anyone found a proper solution yet ?

**Comment by Etheryte** [comment]
Not sure why this was closed, @xiaoyun-studio? Debugging the serverside part of the code is a crucial part of any serious development, the current approach of simply logging errors and showing an error page is **not** sufficient.

**Comment by jetrii** [comment]
Is there an update on this? This is a 100% must have feature for nuxt to be taken seriously IMO.

**Comment by o1lab** [comment]
..bump

**Comment by uptownhr** [comment]
bump bump

**Comment by LancerComet** [comment]
I have posted configuration for Webstorm in issue #2781

**Comment by bingtimren** [comment]
And this seems to relate to webpack transpiling. I tested with the nuxt starter template. In pages/index.vue, if I put "debugger" under "import AppLogo from '~/components/AppLogo.vue'", open Chrome inspector, upon paused, I cannot find "AppLogo" in the local variable but `__WEBPACK_IMPORTED_MODULE_0__components_AppLogo_vue__`. Now if I change the line "import AppLogo from '~/components/AppLogo.vue" to "const AppLogo = require('~/components/AppLogo.vue')" and reload page, upon pause, I can find AppLogo in local variables.

**Comment by bingtimren** [comment]
Ok. For the local variable disappear problem, if I set devtool to 'eval' with the following code in nuxt config.

build: {
    extend (config, { isDev, isClient }) {
      config.devtool = 'eval'
      .....

At break point, I can see the "real" code my browser is running (after compiled), not the code I wrote in the component. The real code would, of course, works exactly the same as the code I wrote (at least, should be), but is different. This should explain the missing local variables.
